tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/logging/TraceValve.java

```diff
@@ -0,0 +1,130 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.runtimes.tomcat.logging;
+
+import com.google.cloud.trace.Trace;
+import com.google.cloud.trace.Tracer;
+import com.google.cloud.trace.core.SpanContext;
+import com.google.cloud.trace.core.TraceContext;
+import com.google.cloud.trace.service.TraceGrpcApiService;
+import com.google.cloud.trace.service.TraceService;
+
+import org.apache.catalina.LifecycleException;
+import org.apache.catalina.connector.Request;
+import org.apache.catalina.connector.Response;
+import org.apache.catalina.valves.ValveBase;
+import org.apache.juli.logging.Log;
+import org.apache.juli.logging.LogFactory;
+
+import java.io.IOException;
+import java.util.Optional;
+import javax.servlet.ServletException;
+
+/**
+ * This valve sends information about the requests to the Stackdriver Trace service.
+ */
+public class TraceValve extends ValveBase {
+
+  /**
+   * Header used by GCP to stores the trace id and the span id.
+   */
+  private static final String X_CLOUD_TRACE_HEADER = "x-cloud-trace-context";
+
+  private static final Log log = LogFactory.getLog(TraceValve.class);
+
+  private TraceService traceService;
+
+  /**
+   * Delay in second before the trace scheduler send the traces (allow buffering or traces).
+   */
+  private Integer traceDelay;
+
+  /**
+   * Identifier of the project to use when sending traces.
+   */
+  private String projectId;
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Initialize the Trace service.</p>
+   */
+  @Override
+  protected void initInternal() throws LifecycleException {
+    super.initInternal();
+
+    if (traceDelay == 0) {
+      throw new LifecycleException("The delay for trace must be greater than 0");
+    }
+
+    if (projectId.isEmpty()) {
+      throw new LifecycleException("You must specify a project to store the traces");
+    }
+
+    try {
+      traceService = TraceGrpcApiService.builder()
+          .setScheduledDelay(traceDelay)
+          .setProjectId(projectId)
+          .build();
+
+      Trace.init(traceService);
+      log.info("Trace service initialized for project: " + projectId);
+    } catch (IOException e) {
+      log.error("An error occurred during the initialization of the Trace valve", e);
+      throw new LifecycleException(e);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Create a new trace containing information about the requests.
+   *    The traces are buffered by the TraceService and regularly sent to Stackdriver</p>
+   */
+  @Override
+  public void invoke(Request request, Response response) throws IOException, ServletException {
+    Tracer tracer = traceService.getTracer();
+
+    Optional.ofNullable(request.getHeader(X_CLOUD_TRACE_HEADER)).ifPresent(traceHeader -> {
```

Comment:
I don't think it makes sense to use `Optional` here - a simple `if (request.getHeader(X_CLOUD_TRACE_HEADER) == null)` works the same and is more readable.